### PR TITLE
Improve Order Flow: Add Required Volume Validation #3722

### PR DESCRIPTION
**Pull request by RasmusBurge-CG**
## Description

This PR introduces validation to ensure the required volume is entered for containerised samples.

#### Case Sample

- New validation method: `validate_case_required_volume`
- New error: `VolumeRequiredCaseError`
- New utility: `is_required_sample_volume_invalid`
- New tests for `validate_case_required_volume`

#### Sample

- New validation method: `validate_required_volume`
- New error: `VolumeRequiredError`
- New utility: `is_required_volume_invalid`
- New tests for `validate_required_volume`

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: tests/services/order_validation_service/sample_rules/test_sample_rules.py
Comment (islean) [inline tests/services/order_validation_service/sample_rules/test_sample_rules.py:117]:
```suggestion
    # THEN no error should be returned
```

File: cg/services/order_validation_service/errors/case_sample_errors.py
Comment (islean) [inline cg/services/order_validation_service/errors/case_sample_errors.py:128]:
```suggestion
class VolumeRequiredError(CaseSampleError):
```
In general we have decided that the errors are cleaner and more readable if we do not include the "CaseSample" etc in the name.

File: cg/services/order_validation_service/errors/sample_errors.py
Comment (islean) [reply cg/services/order_validation_service/errors/sample_errors.py:51]: It would also be a bit more harmonious if the Sample error and the CaseSample errors had the same error message.

File: tests/services/order_validation_service/sample_rules/test_sample_rules.py
Comment (islean) [inline tests/services/order_validation_service/sample_rules/test_sample_rules.py:110]:
```suggestion
    # GIVEN an order with a sample not in a container and no volume set
```

File: tests/services/order_validation_service/test_case_sample_rules.py
Comment (islean) [inline tests/services/order_validation_service/test_case_sample_rules.py:55]:
```suggestion
def test_missing_volume(valid_order: OrderWithCases):
```

File: tests/services/order_validation_service/test_case_sample_rules.py
Comment (islean) [inline tests/services/order_validation_service/test_case_sample_rules.py:57]:
```suggestion
    # GIVEN an orders with two samples with missing volumes
```

File: tests/services/order_validation_service/test_case_sample_rules.py
Comment (islean) [inline tests/services/order_validation_service/test_case_sample_rules.py:64]:
```suggestion
    # WHEN validating that required volumes are set
```

File: tests/services/order_validation_service/test_case_sample_rules.py
Comment (islean) [inline tests/services/order_validation_service/test_case_sample_rules.py:70]:
```suggestion
    # THEN the error should concern the missing volumes
```

File: tests/services/order_validation_service/test_case_sample_rules.py
Comment (islean) [inline tests/services/order_validation_service/test_case_sample_rules.py:78]: This test is likely not needed since there is an integration test for a valid TomteOrder which is an OrderWithCases.

File: tests/services/order_validation_service/sample_rules/test_sample_rules.py
Comment (islean) [reply tests/services/order_validation_service/sample_rules/test_sample_rules.py:99]: Also the samples already have valid volumes in the fixture (or at least they should have).

File: tests/services/order_validation_service/test_case_sample_rules.py
Comment (islean) [inline tests/services/order_validation_service/test_case_sample_rules.py:94]:
```suggestion
def test_missing_volume_no_container(valid_order: OrderWithCases):
```

File: tests/services/order_validation_service/test_case_sample_rules.py
Comment (islean) [inline tests/services/order_validation_service/test_case_sample_rules.py:96]:
```suggestion
    # GIVEN an order with a sample with missing volume, but which is in no container
```

File: tests/services/order_validation_service/test_case_sample_rules.py
Comment (islean) [inline tests/services/order_validation_service/test_case_sample_rules.py:100]:
```suggestion
    # WHEN validating that the order has required volumes set
```

**Review by islean — COMMENTED** [review]
Quite a nitpicky review but in general I write comments about everything I see, and it is up for you to agree or not. See if you agree with my comments and if you want to discuss them I am always available!

File: cg/services/order_validation_service/errors/sample_errors.py
Comment (RasmusBurge-CG) [reply cg/services/order_validation_service/errors/sample_errors.py:51]: Sure, let's keep it simple: `Volume is required`. Yes, I agree that using the same error message makes sense. That was an unintentional mistake on my part.
```suggestion
        f"Volume is required"
```

File: cg/services/order_validation_service/rules/case_sample/utils.py
Comment (RasmusBurge-CG) [reply cg/services/order_validation_service/rules/case_sample/utils.py:64]: Aha! Perfect, that’s much clearer. I'll make a new commit with that.

File: cg/services/order_validation_service/rules/case_sample/rules.py
Comment (RasmusBurge-CG) [reply cg/services/order_validation_service/rules/case_sample/rules.py:216]: Sure, I agree, it does become a bit unclear.

File: tests/services/order_validation_service/test_case_sample_rules.py
Comment (RasmusBurge-CG) [reply tests/services/order_validation_service/test_case_sample_rules.py:78]: Hmm, so I wouldn’t need to test it at all, or should I use a different test set?

Comment (RasmusBurge-CG) [reply tests/services/order_validation_service/test_case_sample_rules.py:78]: Hopefully I’m not spamming you with too many messages, but could you @islean help me understand this?

Comment (islean) [reply tests/services/order_validation_service/test_case_sample_rules.py:78]: Exactly - I think we should have one test which performs all validations for whole valid order. Then we do not need tests which make sure that each individual validation passes for a valid order.

Might need to set up the order a bit intricately to test all the caveats but in principle I think it is a better approach. This test won't cause any issues but adding a "valid order" test for each rule might make the test files a bit rough to read.

Comment (RasmusBurge-CG) [reply tests/services/order_validation_service/test_case_sample_rules.py:78]: Yes, that makes so much sense—I hadn’t thought of it that way.

File: cg/services/order_validation_service/errors/sample_errors.py
Comment (diitaz93) [inline cg/services/order_validation_service/errors/sample_errors.py:52]: make this one line

File: tests/services/order_validation_service/sample_rules/test_sample_rules.py
Comment (diitaz93) [inline tests/services/order_validation_service/sample_rules/test_sample_rules.py:83]: add type hint

File: tests/services/order_validation_service/sample_rules/test_sample_rules.py
Comment (diitaz93) [inline tests/services/order_validation_service/sample_rules/test_sample_rules.py:115]: add type hint

**Review by diitaz93 — COMMENTED** [review]
Looks good! I lack the branch knowledge so Isak has the final word for the approval 😉

File: tests/services/order_validation_service/test_case_sample_rules.py
Comment (RasmusBurge-CG) [reply tests/services/order_validation_service/test_case_sample_rules.py]: I created a new test that validates all the current rules against a case with valid values. Please let me know what you think.

Comment (islean) [reply tests/services/order_validation_service/test_case_sample_rules.py]: Sorry, I was a bit unclear in what I meant. There is already such a test for Tomte, check out `test_valid_order(valid_order: TomteOrder, tomte_validation_service: TomteValidationService)` in `test_validation_integration.py` file. We have a method in all `OrderValidationServices` called `_get_errors` which runs all the validation rules for the specified workflow and returns any errors. We should probably have one per workflow but just by setting your rules to be included in the Tomte validation rules, this is already taken care of.

File: cg/services/order_validation_service/rules/case_sample/utils.py
Comment (islean) [inline cg/services/order_validation_service/rules/case_sample/utils.py:62]: We could move this to cg/services/order_validation_service/rules/utils.py so as to remove some duplication

File: cg/services/order_validation_service/rules/case_sample/utils.py
Comment (islean) [inline cg/services/order_validation_service/rules/case_sample/utils.py:63]:
```suggestion
    """Check if a sample is missing its volume."""
```

File: cg/services/order_validation_service/rules/case_sample/utils.py
Comment (RasmusBurge-CG) [reply cg/services/order_validation_service/rules/case_sample/utils.py:62]: Makes sense!

**Comment by RasmusBurge-CG** [comment]
I'll merge it when I get the green light :)

**Review by islean — APPROVED** [review]
Looks super! ⭐

**Pull request merged**
